src/history/hs_rec.c

```diff
@@ -669,6 +669,26 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
                 F_SET(first_non_ts_upd, WT_UPDATE_CLEARED_HS);
             }
 
+            /* We should never write a prepared update to the history store. */
+            WT_ASSERT(session,
+              upd->prepare_state != WT_PREPARE_INPROGRESS &&
+                upd->prepare_state != WT_PREPARE_LOCKED);
+
+            /*
+             * Skip the check if we don't have a snapshot. Therefore, the check doesn't ensure that
+             * we never write uncommitted updates to the history store. But this is the best we can
+             * do for now.
+             *
+             * Sometimes the application and the checkpoint threads will fall behind the eviction
+             * threads, and they may choose an invisible update if the update was previously
+             * selected by a failed eviction pass. In that case, there is no point in checking the
+             * visibility of such history store updates.
+             */
+            WT_ASSERT(session,
+              !F_ISSET(session->txn, WT_TXN_HAS_SNAPSHOT) ||
```

Comment:
can we check this case irrespective of whether transaction has a snapshot or not. In other words, can we have a utility function which returns whether a transaction is committed or not irrespective of caller transaction state and use that here.

Reply:
I'm not sure if that's viable and I'll have a try.